litt/util/file_lock.go

```diff
@@ -0,0 +1,210 @@
+	logger logging.Logger
+	path   string
+	file   *os.File
+}
+
+// IsProcessAlive checks if a process with the given PID is still running
+func IsProcessAlive(pid int) bool {
+	if pid <= 0 {
+		return false
+	}
+
+	// Send signal 0 to check if process exists
+	// This doesn't actually send a signal, just checks if we can send one
+	err := syscall.Kill(pid, 0)
+	if err == nil {
+		return true
+	}
+
+	// Check the specific error
+	var errno syscall.Errno
+	if errors.As(err, &errno) {
+		switch {
+		case errors.Is(errno, syscall.ESRCH):
+			// No such process
+			return false
+		case errors.Is(errno, syscall.EPERM):
+			// Permission denied, but process exists
+			return true
+		default:
+			// Other error, assume process exists to be safe
+			return true
+		}
+	}
+
+	// Unknown error, assume process exists to be safe
+	return true
+}
+
+// parseLockFile parses a lock file and returns the PID if valid
+func parseLockFile(path string) (int, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read lock file: %w", err)
+	}
+
+	lines := strings.Split(string(content), "\n")
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if strings.HasPrefix(line, "PID: ") {
+			pidStr := strings.TrimPrefix(line, "PID: ")
+			pid, err := strconv.Atoi(pidStr)
+			if err != nil {
+				return 0, fmt.Errorf("invalid PID in lock file: %s", pidStr)
+			}
+			return pid, nil
+		}
+	}
+
+	return 0, fmt.Errorf("no PID found in lock file")
+}
+
+// NewFileLock attempts to create a lock file at the specified path. Fails if another process has already created a
+// lock file. Useful for situations where a process wants to hold a mutual exclusion lock on a resource.
+// The caller is responsible for calling Release() to release the lock.
+func NewFileLock(logger logging.Logger, path string, fsync bool) (*FileLock, error) {
+	path, err := SanitizePath(path)
+	if err != nil {
+		return nil, fmt.Errorf("sanitize path failed: %v", err)
+	}
+
+	// Try to create the lock file exclusively (O_EXCL ensures it fails if file exists)
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+	if err != nil {
+		if os.IsExist(err) {
+			// Lock file exists, check if it's stale
+			if pid, parseErr := parseLockFile(path); parseErr == nil {
+				if !IsProcessAlive(pid) {
+					// Process is dead, remove stale lock file and try again
+					if removeErr := os.Remove(path); removeErr != nil {
+						return nil, fmt.Errorf("failed to remove stale lock file %s: %w", path, removeErr)
+					}
+
+					// Try to create the lock file again
+					file, err = os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+					if err != nil {
+						return nil, fmt.Errorf("failed to create lock file after removing stale lock %s: %w",
+							path, err)
+					}
+				} else {
+					// Process is still alive, cannot acquire lock
+					debugInfo := ""
+					if content, readErr := os.ReadFile(path); readErr == nil {
```

Comment:
if ReadFile fails here, would it make sense to note the failure in `debugInfo`? (also below)